dfir-iris/iris-web
# Rest improvement

Start of implementation of REST API v2. Closes #502. Closes #503.
Contains start of #519 and #534.

Main things that were done:
* introduction of new endpoints, deprecation of previous endpoints (see below for details)
* deprecated endpoints were replaced in IHM
* addition of tests for new REST endpoints
* moved all rest endpoints into namespace app.blueprints.rest
* split tests into tests_rest and tests_graphql
* removed annotation `ac_api_case_requires`
* added the access levels to check as parameter to `ac_requires_case_identifier` 
* several code cleanups
* renamed endpoint_deprecated into endpoint_removed and added new endpoint_deprecated which does not return an error, but still execute the code (for a graceful migration)
* added file app.blueprint.rest.endpoints with response_api_success, response_api_deleted, response_api_created, reponse_api_error, response_api_not_found for the new API return codes
* updated version of setuptools 70.3.0

New endpoints API
* POST /api/v2/cases
* GET /api/v2/cases
* GET /api/v2/cases/{identifier}
* DELETE /api/v2/cases/{identifier}
* POST /api/v2/cases/{identifier}/iocs
* GET /api/v2/iocs/{identifier}
* DELETE /api/v2/iocs/{identifier}
* POST /api/v2/cases/{identifier}/tasks
* GET /api/v2/tasks/{identifier}
* DELETE /api/v2/tasks/{identifier}

Deprecated endpoints
* POST /manage/cases/add
* GET /manage/cases/{identifier}
* GET /case/exists
* DELETE /manage/cases/delete/{identifier}
* POST /case/ioc/add
* GET /case/ioc/{identifier}
* GET /case/ioc/state
* DELETE /case/ioc/delete{identifier}
* POST /case/tasks/add
* GET /case/tasks/{identifier}
* DELETE /case/task/delete/{identifier}

Conversions of endpoints in interface
* POST /manage/cases/add => POST /api/v2/cases
* GET /manage/cases/list => GET /api/v2/cases
* GET /manage/cases/filter => GET /api/v2/cases
* GET /manage/cases/{identifier} => GET /api/v2/cases/{identifier}
* GET /case/exists => GET /api/v2/cases/{identifier}
* DELETE /manage/cases/delete/{identifier} => DELETE /api/v2/cases/{identifier}
* POST /case/ioc/add => POST /api/v2/cases/{identifier}/iocs
* GET /case/ioc/{identifier} => GET /api/v2/iocs/{identifier}
* GET /case/ioc/state => GET /api/v2/iocs/{identifier}
* DELETE /case/ioc/delete{identifier} => DELETE /api/v2/iocs/{identifier}
* POST /case/tasks/add => POST /api/v2/tasks
* GET /case/tasks/{identifier} => GET /api/v2/tasks/{identifier}
* DELETE /case/task/delete/{identifier} => DELETE /api/v2/tasks/{identifier}

Warning: the type when retrieving cases (GET /api/v2/cases) is not the same as the type returned by (GET /api/v2/cases/{identifier}) and (POST /api/v2/cases). They should (and will at some point) be aligned. This may be a problem if the tool is released before they are aligned: we need to talk about the API evolution strategy (either have new endpoints: /api/v3/cases, or release this API in beta, but then maybe we shouldn't have already deprecated the previous endpoints, other?).
